pennylane/grouping/utils.py

```diff
@@ -0,0 +1,492 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Utility functions used in Pauli partitioning and measurement reduction schemes utilizing the
+symplectic vector space representation of Pauli words.
+"""
+
+from pennylane import PauliX, PauliY, PauliZ, Identity
+from pennylane.operation import Observable, Tensor
+import numpy as np
+
+
+def get_n_qubits(observables):
```

Comment:
Hey @ral9000, great work! Let me try to explain why the tests may fail here, which is part of a very recent, cross-cutting refactor of PennyLane's wire management system...

Since v0.11.0, PL has the new feature of letting users specify their own wire labels (by optionally creating a device with ``device('dev.name', wires=['a', 'x', 'q1'])`` instead of ``device('dev.name', wires=3)``). The devices translate the wire labels to their own labels, for example to consecutive integers ``0, 1,...``.

We realised during the refactor that up to then PL depended in many places on the fact that wire labels are consecutive integers and by that reveal information on *where* wires are in a curcuit, as in your `get_n_qubits` function - you assume that you can conclude the number of qubits and the order of the observables acting on them (as well as intermittent identities) from the wire labels. 

On a very deep level, we realised that this would however be the wrong abstraction - an observable is an operator that might know the label of the wire it gets applied to, but it should not contain information on a circuit and order of qubits if we do not want to mix up conceptually separate parts. For example, an observable ``Obs(wires=['b'])`` is well defined without fixing the fact that ``'b'`` is always the second wire, as it could be applied in a circuit with wires ``['a', 'b']`` and in another circuit with wires ``['c', 'd', 'b']``.

This improved abstraction however requires some extra work in utility functions that process observables in the context of circuits. These now need to know the absolute "wire register" with the custom wire labels of all qubits in order to infer the order.  

Cutting a long story short, the way this was solved in the qchem module is to have an optional argument ``wires`` where the user can feed in the custom wire labels like ``wires = ['a', 'x', 'q1']``. If none are given, we assume that the user does not make use of custom labels and set ``wires=range(num_wires)`` (if the total number of wires is known), or - like you do here - infer the number of wires as needed.

Does this make sense? 

I'll need have a closer look at the module to see if this solution works here...

Reply:
Hi @mariaschuld, thanks a lot! I meant to check the PR still worked with the latest internal changes but made a silly Git mistake. Thanks for outlining the changes to specifying wires. Looking at the current `Wires` class, it seems  `get_n_qubits` can be simplified drastically using the `Wires`'s static method `all_wires`, i.e.

`n_qubits = len(Wires.all_wires([obs.wires for obs in observables]))`

Then this should return the minimum number of qubits even if abstract labels are specified. I'll push the necessary changes to this PR.

Reply:
Thanks @ral9000; I've temporarily added the `WIP` label until the changes are in

Reply:
I might have confused myself, does `get_n_qubits` only get the minimum number of qubits required so that there are enough unique qubits for every operator to act on? For example, for `[MyObs(wires='1'), MyObs(wires=['1', '13'])]` the number of qubits is 2? Then my lecture was not really necessary here :) 

I somehow thought you want to infer the maximum number of qubits from the integer label, i.e. in the above example `13`.

Anyhow, let us know when the latest changes are ready :)

Reply:
Thanks @josh146 !

@mariaschuld ,

Some confusion is definitely warranted -- Initially, I indeed wanted `get_n_qubits` to infer the number of qubits from the highest integer label. Now that I see the `wires` attribute is moreso a set of abstract labels, this inference is somewhat limited. I've now changed `get_n_qubits` to simply check the size of the non-trivial support of the observable, so now `get_n_qubits([MyObs(wires=[1, 13])])` = `get_n_qubits([MyObs(wires=['a', 'b'])])` = 2.

Reply:
Thanks for pointing this out @antalszava , originally `get_n_qubits` was used in a few places, like `convert_observables_to_binary`, for obtaining the smallest dimensionality possible when mapping a set of Pauli words to the binary representation. It seems obsolete now, given that the changes to the `Wires` class let you obtain this information in one line, and can be removed.